src/features/code_actions.zig

```diff
@@ -93,6 +118,104 @@ pub const Builder = struct {
     }
 };
 
+pub fn generateStringLiteralCodeActions(
+    builder: *Builder,
+    token: Ast.TokenIndex,
+    actions: *std.ArrayListUnmanaged(types.CodeAction),
+) !void {
+    const tags = builder.handle.tree.tokens.items(.tag);
+    switch (tags[token -| 1]) {
+        // Not covered by position context
+        .keyword_test, .keyword_extern => return,
+        else => {},
+    }
+
+    const token_text = offsets.tokenToSlice(builder.handle.tree, token); // Includes quotes
+    const parsed = std.zig.string_literal.parseAlloc(builder.arena, token_text) catch |err| switch (err) {
+        error.InvalidLiteral => return,
+        else => |other| return other,
+    };
+    // Check for disallowed characters
+    for (parsed) |c| {
+        switch (c) {
+            0x01...0x09, 0x0b...0x0c, 0x0e...0x1f, 0x7f => return,
+            else => {},
+        }
```

Comment:
Something like this should be more readable:
```suggestion
        if (c == '\n' or c == '\r') continue;
        if (std.ascii.isControl(c)) return;
```

Also, read my final review comment about carriage returns.